### ISSUE TITLE
rule suggestion - conditional expression

### ISSUE THREAD

**Issue by aervin** [fimbullinter/wotan#221]
I'm proposing a easily configurable conditional expression rule which goes a bit farther than the existing TSLint rule in that it should also catch return statements and call expressions.

Rather than a config which relies on bool flags for each case (assignment/return/call), I'd like there to be "modes": "high", "medium, "low", "disallow".

"high mode":
* Catches all if statements which could be transformed into conditional expressions--even ifs w/ nested ifs and multiline blocks.

"medium mode":
* Ignores call expressions and nested ifs

"low mode":
* Would essentially be what the existing TSLint rule provides.

"disallow":
* I guess this would be pretty easy to implement. Not a style I love, but...

**Comment by ajafff** [comment]
I like the idea. Just to make sure I completely understand your proposal:

> it should also catch return statements and call expressions.

"return statements" is pretty clear, you mean something like this:

```ts
if (Boolean()) {
    return 'a';
} else {
    return 'b';
}
// or
if (Boolean()) {
    return 'a';
}
return 'b';

// fixed to
return Boolean() ? 'a' : 'b';
```

But what exactly does "call expressions" mean?

```ts
// different functions to call?
if (Boolean()) {
  foo();
} else {
  bar();
}
// won't work for methods or incompatible signatures
if (Boolean()) {
  this.foo();
} else {
  this.bar();
}

// or is it different argument?
if (Boolean()) {
  doStuff('foo');
} else {
  doStuff('bar');
}

// won't always work for functions with overloads
declare function doStuff(arg: string): string;
declare function doStuff(arg: number): number;
if (Boolean()) {
  doStuff('foo');
} else {
  doStuff(1);
}
// this will give a type error
doStuff(Boolean() ? 'foo' : 1);
```

**Comment by aervin** [comment]
Isn't this valid @ajafff:
```ts
if (true) {
  funcA();
} else {
  funcB();
}

// could be
true ? funcA() : funcB();
```

**Comment by ajafff** [comment]
@aervin you are absolutely right. My thoughts were way too complicated...

That means it's not only limited to call expressions. The rule can handle every IfStatement that has exactly one ExpressionStatement (or ReturnStatement) in each branch.

**Comment by aervin** [comment]
I never used the `check-else-if` setting for TSLint's rule--do we need an equivalent here? Unless you use it, I'd suggest we hold off until the functionality is requested.

**Comment by ajafff** [comment]
Just a heads up: I'm trying to avoid rules about code style in the set of core rules. I'll add another package to this repo for such rules.
This rule will be a better fit for that new package as there is no sensible default that satisfies almost every user.
Another rule that will most likely move to the other package is `parameter-properties`.

Though that shouldn't hinder any implementation effort. Moving the rule afterwards is not a big deal.

**Comment by aervin** [comment]
Sounds sensible to me. 

I've got a good start on the "high" setting. However, it currently misses this:
```ts
if (Boolean()) {
    return 'a';
}
return 'b';
```
and does not check else-ifs. 

I'll open a PR once I have a more complete implementation.

**Comment by ajafff** [comment]
Regarding the IfStatements with early return: I have a [work-in-progress branch](https://github.com/fimbullinter/wotan/commit/212c1d315db04b86b54342d8a84addfd8851d1da) or a a rule that detects a similar pattern. Maybe you can reuse some of the logic.

```ts
if (Boolean()) {
  v = true;
} else {
  v = false;
}

if (Boolean()) {
  return false;
}
return true;
```

**Comment by aervin** [comment]
Thanks @ajafff I will definitely take a look.

Before I forget, I was wondering what's involved for creating "default" configs (like `wotan:recommended`). For the style package, I can imagine multiple useful defaults such as `less-code` and `beginner-friendly`.

**Comment by ajafff** [comment]
I really like the idea to add presets like `beginner-friendly` and `less-code`.

Creating config presets is easy: just create a `<name>.yaml` file in the root of the package.

Let's assume that new package is called `@fimbul/codestyle`.
The default config (`index.yaml`) of a rules package should not contain any enabled rules. Is should just expose the rulesDirectory to extending configs:

```yaml
# packages/codestyle/index.yaml
rulesDirectories:
  codestyle: src/rules
```

```yaml
# consumer.yaml
extends: "@fimbul/codestyle" # nothing is enabled b default
rules:
  codestyle/conditional-expression: error
```

Let's add a preset with some rules enabled:

```yaml
# packages/codestyle/beginner-friendly.yaml
extends: ./index.yaml
rules:
  codestyle/conditional-expression:
    options:
      mode: never
  codestyle/parameter-properties:
    options:
      mode: never
```

```yaml
# consumer.yaml
extends: "@fimbul/codestyle/beginner-friendly"
rules: #override some rules if necessary
  codestyle/parameter-properties:
    options:
      mode: consistent
```

**Comment by aervin** [comment]
This could not be more straightforward. Well done @ajafff
